### DISCUSSION TITLE
Regarding the deprecation and ultimately removal of `--initialize-at-build-time=`

### DISCUSSION THREAD

**Discussion by zakkak** [oracle/graal#3476] [Q&A] [answered]
Hello,

I would like to ask about the justification behind the deprecation and ultimately removal of `--initialize-at-build-time=` in 22.0 (see https://github.com/oracle/graal/commit/5013ca7c9d88bd12de2ebd24e3196a40d9044c9c)

Is this to prohibit the use of the flag (in that specific way) in order to prevent issues like https://github.com/oracle/graal/pull/3417?
Or is it based on something new coming in like an improved analysis that will result in more aggressive, yet reliable, compile-time initialization? 

Thanks

cc @vjovanov

**Comment by borkdude** [comment, marked as answer]
I also would like to know more about this. Clojure projects rely on this feature. As far as a I know, it is _impossible_ to build a Clojure project without this option. I asked about this to @shelajev and he told me to ping @vjovanov about this. Should we be concerned that 21 is the last version that will support Clojure projects (like babashka and clj-kondo, which have grown quite popular over time), or does the warning just mean that there will be another way to accomplish the same behavior that we got with this flag in 21, etc?

I'm willing to "sit down" with one of the GraalVM devs and walk through what is needed to compile a very simple Clojure project and what errors you see when not using the `--initialize-at-build-time` option.

**Reply by ericdallo** [reply]
Good point, [clojure-lsp](https://clojure-lsp.github.io/clojure-lsp/) heavily relies on that as well

**Reply by thomaswue** [reply]
Hi @borkdude! Thanks for your comment. Note that this is not about removing the option, but rather requiring to list the classes/packages that should be initialized at build time explicitely. Is this not possible for Clojure projects? Could there be some less general wildcard like "org.clojure.*" or is a fully general wildcard ".*" required?

**Reply by borkdude** [reply]
@thomaswue For clojure projects, I expect the list of packages to be quite long (like: every Clojure namespace that is compiled basically, not only of Clojure itself, but also from the user). So if there would be a wildcard option like `**` to select everything which can then be overridden with `--initialize-at-run-time=...`, this would work.

So if `--initialize-at-build-time=**` (or a similar notation) is going to do what `--initialize-at-build-time` is doing now, then I don't see any problems.

**Reply by fcurts** [reply]
`--initialize-at-build-time` works just fine for our Truffle language even though we use many third-party packages. It seems that having to maintain a list of included packages would complicate matters for no gain. Is there any reason to _not_ support a general wildcard?

**Reply by thomaswue** [reply]
How are those libraries referenced, is it via maven artefacts? Is there a way for those libraries to declare themselves as "build time initialization required"?

Our main aim is that framework authors should be explicit about whether build time initialization should be used for a certain library.

**Reply by thomaswue** [reply]
@fcurts Does it also work with the automatic detection when nothing is specified or are those libraries *required* to be initialized at build time?

**Reply by borkdude** [reply]
@thomaswue I tried `--initialize-at-build-time=.` and this seems to work for Clojure projects, so if GraalVM 22 keeps supporting this, all is well from the Clojure side.

**Reply by adinn** [reply]
> Our main aim is that framework authors should be explicit about whether build time initialization should be used for a certain library.

This is the critical point. Build-time initialization (BTI) of classes has the potential to build invalid 'build-time' state into the runtime. How that might happen can often be very subtle because it depends on how static init of data and later use of data are related. That relationship can be arbitrarily complex since it involves execution of whatever unconstrained static init code a framework or app chooses to implement.

If an end user tries to force BTI for framework code (or even JDK runtime code) they may well not be in a position to recognize such invalid usage and understand the consequences. If you apply BTI arbitrarily to classes without actually understanding how they are designed your code may well not operate as expected. This does not just have the potential for breakage. A worse outcome can be subtly different behaviour compared to how the app runs on the JVM. That could lead to functional, data integrity and security errors.

Now, of course, GraalVM includes a host of checks to ensure that BTI is not incorrectly applied to JDK runtime code and there is every reason to believe that this is operating correctly. Similarly, middleware developers have worked hard to ensure that the same applies to middleware libraries. However, it is not possible for it to guard against every potential invalid BTI for all possible JDK, middleware or end user code and certainly is not reliable if end users start driving the selection of classes for BTI without detailed knowledge of the code they are configuring for BTI.

Hopefully it will be possible to improve detection of invalid BTI incrementally over time and, perhaps, even to upgrade JDK, framework and app implementations and/or the Java language itself to help do so. However, at present the only way to ensure that BTI is safe is for those who design and build the code to identify which components of their code base should use BTI. So, shouting hurrah and carrying on regardless here simply because you get a free pass via some sort of wildcard exemption on this command line flag is seriously missing the point. if you want to use BTI widely you need to know what you are doing. Going ahead and doing it anyway without knowing what you are doing just because some syntax is available to do so is not a good enough reason.

**Reply by borkdude** [reply]
@adinn I agree that it is good to make GraalVM users aware of the consequences of initializing at built time by default. 

However, taking away this option would lead to breakage of already many existing Clojure projects that leverage GraalVM native-image and are fully functional (in so far that their test suites demonstrate that they are working correctly). 

Detection of erroneous built-time initialization is often figured out on a library level and documented, like here:
https://github.com/http-kit/http-kit#class-initialization.

Having the blanket option available via the syntax is imo the right thing to do, instead of causing a disruptive change. Print warnings all you want, that is fine to make users aware, but this is still better than taking away something which will cause projects to no longer function at all.

**Reply by vjovanov** [reply]
@borkdude the fact that frameworks and libraries use this option is exactly the reason we have to remove it. This means that these libraries won't work with say AWT--not the best thing to break the code composition of a programming language. 

If you think 22.0.0 is too early, this can be postponed, but a deadline is necessary or this option creates non-composable fractions of the JVM ecosystem.

**Reply by borkdude** [reply]
@vjovanov As long as there is a way for the end user (as opposed to libraries or frameworks) to accomplish this, I don't see a problem for Clojure applications. A deadline is not the point, it's more like offering a good alternative for existing projects. I mean, I don't see any alternative for Clojure projects, no matter what the deadline is.

**Reply by thomaswue** [reply]
@borkdude Would it be possible to specify this as part of the maven definition when a project dependency is added?

Our concern basically is that a user would add a dependency to a different library that is not designed for init-everything-at-build-time. It is much more likely for a library to work out-of-the-box without this global flag.

**Reply by borkdude** [reply]
@thomaswue I'm not sure I understand what you mean. I usually provide this flag when invoking `native-image` in a shell script.

**Reply by thomaswue** [reply]
How do you specify the Java libraries that end up in the image? Like how does your classpath look like?

**Reply by borkdude** [reply]
@thomaswue Usually I create an uberjar first and use the `-jar` option, but you can also provide the classpath using `clojure` or `lein` which are the main build tools in the Clojure ecosystem. `lein` uses the maven ecosystem as does `clojure`, but the latter can also use git dependencies which are not based on maven (but usually there are at minimum a few maven artifacts on the classpath).

**Reply by borkdude** [reply]
I can give you an example uberjar, if you want...

**Reply by thomaswue** [reply]
Specifying the option for all classes in a specific jar file seems quite reasonable. Would it be OK for this to only work in such broad manner if an uber jar is created first or is that too limiting?

**Reply by borkdude** [reply]
@thomaswue That would work for me.

**Reply by borkdude** [reply]
@thomaswue I was able to come up with a workaround for Clojure.

Using something like:

``` clojure
(->> (map ns-name (all-ns)) (remove #(str/starts-with? % "clojure")) (map #(str/split (str %) #"\.")) (keep butlast) (map #(str/join "." %)) distinct (map munge) (cons "clojure"))
```

you can get a list of all Clojure namespaces (and "packages" in Java terminology) compiled by Clojure.

This can then be used to feed into the `--initialize-at-build-time` setting. I could get rid of a setting for Postgres (but at the same time I needed to add one for jackson, since some of the classes are used at build time from a Clojure namespace).

``` clojure
args=( "-jar" "$BABASHKA_JAR"
       "-H:Name=$BABASHKA_BINARY"
       "-H:+ReportExceptionStackTraces"
       "-J-Dborkdude.dynaload.aot=true"
       "-H:IncludeResources=BABASHKA_VERSION"
       "-H:IncludeResources=SCI_VERSION"
       "-H:ReflectionConfigurationFiles=reflection.json"
       "--initialize-at-build-time=clojure,sci.impl,selmer,babashka.nrepl,babashka.impl.clojure.java,babashka.impl,rewrite_clj.node,bencode,rewrite_clj.parser,babashka.impl.clojure,org.httpkit,rewrite_clj.custom_zipper,rewrite_clj.zip,borkdude.graal,babashka.nrepl.impl,babashka.pods,cognitect,babashka,edamame.impl,cheshire,rewrite_clj,hiccup,sci,borkdude,flatland.ordered,babashka.pods.impl,clj_yaml,babashka.impl.clojure.core,datascript,hf.depstar,babashka.impl.tools,sci.addons,babashka.impl.clojure.test"
       "--initialize-at-build-time=com.fasterxml.jackson"
       "-H:Log=registerResource:"
       "-H:EnableURLProtocols=http,https,jar"
       "--enable-all-security-services"
       "-H:+JNI"
       "--verbose"
       "--no-fallback"
       "--no-server"
       "--report-unsupported-elements-at-runtime"
       # "--initialize-at-run-time=org.postgresql.sspi.SSPIClient"
       "--initialize-at-run-time=org.httpkit.client.ClientSslEngineFactory\$SSLHolder"
       "--native-image-info"
       "--verbose"
       "-H:ServiceLoaderFeatureExcludeServices=javax.sound.sampled.spi.AudioFileReader"
       "-H:ServiceLoaderFeatureExcludeServices=javax.sound.midi.spi.MidiFileReader"
       "-H:ServiceLoaderFeatureExcludeServices=javax.sound.sampled.spi.MixerProvider"
       "-H:ServiceLoaderFeatureExcludeServices=javax.sound.sampled.spi.FormatConversionProvider"
       "-H:ServiceLoaderFeatureExcludeServices=javax.sound.sampled.spi.AudioFileWriter"
       "-H:ServiceLoaderFeatureExcludeServices=javax.sound.midi.spi.MidiDeviceProvider"
       "-H:ServiceLoaderFeatureExcludeServices=javax.sound.midi.spi.SoundbankReader"
       "-H:ServiceLoaderFeatureExcludeServices=javax.sound.midi.spi.MidiFileWriter"
       "$BABASHKA_XMX" )
```

**Reply by thomaswue** [reply]
OK, great! This seems like a more robust solution, because now if you start to reference some random other Java library, it is not forced into build time initialization. I think we should still also make the fat jar option work - i.e., for a specific jar file it should be simple to specify that all the classes in it should be initialized at build time.

**Reply by borkdude** [reply]
Btw, I did some closer investigation and the reason why Clojure programs are prone to initialize-at-build-time is that `clojure.lang.RT` does dynamic class loading in a static initializer:

https://github.com/clojure/clojure/blob/master/src/jvm/clojure/lang/RT.java#L338

There might be more such cases, but this is the first obvious one.

**Reply by borkdude** [reply]
@thomaswue In the [GraalVM docs](https://www.graalvm.org/reference-manual/native-image/ClassInitialization/)

> The goal is to have as many classes as possible initialized at build time, yet keep the correct semantics of the program.

Is this still true or has the pendulum swung into the direction of "as many classes as possible initialized at run time"? What are the trade-offs? Should the docs be updated?

**Reply by thomaswue** [reply]
It is still the case that we are trying to initialize as many classes as possible at build time. An algorithm during the closed world analysis tries to proof as many static constructors as possible as safe for initialization at build time. This succeeds when the static constructor is without side effects. For best performance we also encourage framework developers to enable build time initialization for their libraries. We just don't want to have a *default* to build time initialization without explicit opt-in.

**Reply by fcurts** [reply]
Here's our situation:
* We want all of our Truffle language's dependencies to be initialized at build time.
* _None_ of our OSS dependencies publishes GraalVM metadata.
* The most practical solution for us is to initialize everything at build time and rely on our test suite to find any problems. We are happy to replace dependencies that aren't compatible with build-time initialization, but after using `--initialize-at-build-time` for several years, we have yet to notice a problem. It's worth noting that our dependencies are lower-level than the average Java app's dependencies.

We see `--initialize-at-build-time` as an expert option. If it is removed, we'll need to implement an automated workaround (enumerate all packages at build time, or feed native image a fat Jar).

**Reply by adinn** [reply]
> The most practical solution for us is to initialize everything at build time and rely on our test suite to find any problems. We are happy to replace dependencies that aren't compatible with build-time initialization, but after using --initialize-at-build-time for several years, we have yet to notice a problem. It's worth noting that our dependencies are lower-level than the average Java app's dependencies.

Well, that's what I would call a 'courageous' policy. You might want to take that as a compliment but -- putting on my engineering hat -- I am not sure that you should.

> We see --initialize-at-build-time as an expert option.

Indeed, so do I, which is why I am trying to help identify some way for the developers of library and runtime code, rather than its users, to be the ones who make decisions about how that code should be initialized at build time. I hope you are truly an expert in all that low-level code you are configuring.

It's still your choice, of course -- a-and, transitively, the choice of all those who trust in your code, naturally. Being at the base of that inverted pyramid of trust (in both the OpenJDK and GraalVM projects) you can see why I would want users to consider that choice very carefully.

**Reply by RickMoynihan** [reply]
Hi all,

As a clojurist compiling programs with native-image I've been following this issue with interest.

My, understanding of the issue is that all aot compiled clojure namespaces under native-image require this setting, and that forcing explicit enumeration of them all as bordude has shown would be an additional burden.  I think @thomaswue's uberjar proposal is a better option for clojure workflows; though I do question whether these _must_ be packaged as a jar at all?  Could we not for example just give you a classpath root instead, as not all clojure graal workflows require uberjarring; though they all do require an aot compilation step.  

A small unmentioned background point relevant to clojurists here is that in the clojure community libraries whether distributed on maven or not, are _not_ idiomatically distributed as class files, but as clojure source code instead.  It is then entirely the job of the application to aot compile all applications clojure code into class files.  This is principally done because of the way clojure itself is structured, where aot compiling a library would also mean compiling/linking clojure.core and any transitive dependencies, then disentangling the mess of class files from the library jar.  As far as I know nobody does this outside of perhaps highly controlled builds.

The upshot of this is that essentially every clojure native-image project _will_ need to first aot all the clojure code, with a command equivalent to this: `clojure -M:app -e "(compile 'my_app.main)"` and can control precisely where those class files will be output, the default is to put them in a `classes` folder.  This means _all_ classes derived from clojure code will be in the same folder (separated from any java classes they may also instantiate), and ammenable to a command like `--initialize-at-build-time=classes`.

Is there a way for us to avoid requiring an extra step to supply these as a jar here?  It seems superfluous to me.

**Reply by borkdude** [reply]
@RickMoynihan Small detail: the already compiled classes from Clojure itself (and possibly other AOT-ed Clojure libraries, although this is an anti-pattern) won't end up inside the `classes` directory, so you would still have to provide the classpath `classes:<paths to clojure core jars>` .

**Reply by thomaswue** [reply]
How do the static constructors of the classes output by the clojure compiler typically look like?  It would be indeed interesting to further investigate why those classes need to be marked as build-time initialization.

**Reply by borkdude** [reply]
@thomaswue Don't know if this is useful to you, but you can decompile Clojure expressions fairly easily:

https://gist.github.com/borkdude/eec82865575433c06dadfaa4b50b5c44

But of course you can use `javap` to inspect `.class` files emitted by Clojure as well.

**Reply by borkdude** [reply]
@thomaswue I could be mistaking but how I currently see it is:

1. Before you can do anything useful in a Clojure program you need `clojure.core` loaded. This contains the standard library of functions that all Clojure programs build upon.
2. The loading of `clojure.core` is done in `clojure.lang.RT` in a static initializer here:
https://github.com/clojure/clojure/blob/b1b88dd25373a86e41310a525a21b497799dbbf2/src/jvm/clojure/lang/RT.java#L308-L345
3) Step 2 cannot be done at run time (in native image) because it involves a (dynamic) class loader to load the byte code.
4) Hence everything that relies on `clojure.core` (pretty much all Clojure programs) must be initialized at build time.

**Reply by thomaswue** [reply]
OK, understand. The generated classes all have static fields related to the clojure runtime system that are initialized in static constructors with relatively complex code paths. And I guess this is here not just a performance optimization to run this at build time, but the code paths also exercise some reflection that otherwise would have to be configured?

**Reply by borkdude** [reply]
@thomaswue That too, I think RT uses `Class/forName` to check if the class is already loaded which causes reflection config false positives.

**Reply by borkdude** [reply]
I've created a small project here to get rid of those: https://github.com/borkdude/refl

**Reply by vjovanov** [reply]
@borkdude thanks for pointing this issue out and writing the reflection-cleaning program. We will find a solution for Clojure that works without `--initialize-at-build-time`. We can either use features and substitutions, or we can change how `clojure.lang.RT` works when compiling with Native Image. 

To reduce maintenance overhead, would it be possible to change `clojure.lang.RT` to work differently when compiling with Native Image, e.g., to avoid dynamic class loading?

**Reply by borkdude** [reply]
@vjovanov Perhaps it's easier to patch https://github.com/clojure/clojure/blob/master/src/jvm/clojure/lang/DynamicClassLoader.java with a substitution.

Note that I'm not a member of the core team and I can't decide if and how Clojure will be patched, although I could open an issue, discuss, provide patches and hope for the best.

**Reply by borkdude** [reply]
Btw, here is a talk I did on creating command line binaries with Clojure and GraalVM native-image, it was published today:

https://www.youtube.com/watch?v=L2LAaQBVvxM
Repo that goes with the talk: https://github.com/borkdude/jayfu

**Reply by borkdude** [reply]
One limitation somebody ran into with the above snippet to generate all known packages from the list of available Clojure namespaces was with single segment namespaces, e.g. [this one](https://github.com/ztellman/primitive-math/blob/master/src/primitive_math.clj).
Although it's exceptional and an anti-pattern to have single segment namespaces in libraries, it can happen in practice.
Clojure namespaces like `foo.bar` are compiled into a package `foo` and several classes `bar`,  `bar__init`, etc inside that package.
But when compiling a single segment namespace, the generated classes have no package, e.g. `primitive_math$variadic_predicate_proxy` is one such name. And these classes can't be captured using a package name. Is there a solution for this?

**Reply by vjovanov** [reply]
@borkdude would you have time to meet and discuss the best way to fix this properly, i.e., no external tools etc.

**Reply by borkdude** [reply]
Absolutely.

**Reply by vjovanov** [reply]
We can set it up over the [GraalVM community slack](https://graalvm.slack.com/join/shared_invite/enQtODM1NTQzNDIyMjk0LTNlYjBhNTdiNDQxNTdkNDE0ZmFlNzMwNDcyOWZjOTEzN2UyMTM5MTk5YjY1ZDgwMWU2NGQwNjVhZWVhOWRiOGQ).

**Reply by borkdude** [reply]
After looking into this a bit more, I remain a bit skeptical about the effect on startup time when moving work from static blocks to runtime, even when patching Clojure's dynamic loading, because if you inspect the contents of static initializer blocks of Clojure-generated bytecode (especially the `__init.class` classes), it's non-trivial and not sure if it's worth the endeavor of working around this.  In the meanwhile I think we've found a reasonable workaround by listing Clojure packages explicitly and leaving Java classes untouched.

A reasonable workaround for Clojure classes that should not capture top-level values (or instances) at build time is to use `delay`, like in this commit: https://github.com/clojure/tools.deps.alpha/commit/5f0fb4f23a5e38aaaa1b13165540a63e791aaaf4.

<img width="665" alt="Screenshot 2021-07-01 at 12 31 08" src="https://user-images.githubusercontent.com/284934/124110201-48564780-da68-11eb-8ee5-95e4b979856e.png">
<img width="1086" alt="Screenshot 2021-07-01 at 12 31 20" src="https://user-images.githubusercontent.com/284934/124110240-51471900-da68-11eb-950f-4d6f13e583fb.png">
<img width="597" alt="Screenshot 2021-07-01 at 12 31 28" src="https://user-images.githubusercontent.com/284934/124110279-5a37ea80-da68-11eb-861f-ecb06bc56ba1.png">

**Reply by luontola** [reply]
Starting a Clojure program goes like this:

1. JVM starts loading the `my_app.main` class (which contains the main method).
2. It has a static initializer which tells the Clojure runtime to load the `my-app.main` namespace.
3. Clojure tries to find resources `my_app/main__init.class` and `my_app/main.clj`, and either loads the class file directly or compiles the .clj to .class at runtime.
4. JVM starts loading the `my_app.main__init` class.
5. It has a static initializer for initializing the Clojure namespace. This includes:
    * Load recursively (steps 3-5) all namespaces that the current namespace refers to.
    * Instantiate all functions and add them to the [namespace object](https://github.com/clojure/clojure/blob/master/src/jvm/clojure/lang/Namespace.java) (basically a hashmap from names to functions/objects).
    * Instantiate all fields and add them to the namespace object (the field initializer codes may trigger the loading of many more Java classes).
6. After there are no more namespaces to load, the static initializer in the `my_app.main` class is finished and the JVM can call the main method `my_app.main.main(String[])`

Thus Clojure's code loading works eagerly; it loads on startup every namespace that the program uses.

Due to the dynamic code generation and reflection which happens when loading Clojure code, it's best for native-image to initialize it at build time - likely runtime initialization won't even work. 

If I've understood correctly, native-image's build-time initialization (BTI) requires that a every class which is touched during BTI must also use BTI. The static initializer of a BTI class cannot refer to run-time initialized (RTI) classes.

To make it easier to native compile Clojure apps, it would be nice to have a `--initialize-at-build-time-recursively=my_app.main` flag that would BTI the class(es) given as a parameter **and all other classes which are used during its initialization**. For a Clojure app, that one parameter would recursively BTI all the Clojure code and hundreds of Java classes that were used during the initialization. If a class is explicitly marked for RTI, it would be an error if the recursive BTI reaches it. In such a case, the Clojure code would need to be rewritten to avoid that initialization (e.g. using `clojure.core/delay`).

There's also the matter of native-image-agent. Currently it generates thousands of entries to reflect-config.json and resource-config.json during Clojure's namespace loading. Most of them are not needed after BTI. If `--initialize-at-build-time-recursively` is implemented, it would benefit from a companion parameter in native-image-agent (e.g. `-agentlib:native-image-agent=initialize-before-startup=my_app.main`), so that the agent would first run the static initializer of the class(es) given as parameter, and only after that start monitoring reflection and resource usages. (It should be quite easy to prototype this by writing a Java agent which just loads a class, and ordering the Java agents so that it happens before running native-image-agent. Maybe I'll try it at some point.)

**Reply by borkdude** [reply]
@luontola I agree with the above.

For the upcoming deprecation, I and some others have developed this library:

https://github.com/clj-easy/graal-build-time

It will look for packages defined by Clojure (by finding classes ending with `__init.class`) and automatically will flag those (and not more than those) packages for BTI. All you have to do is to include this library in your uberjar or classpath while running `native-image`.

**Reply by borkdude** [reply]
@luontola Another useful project may be this one: https://github.com/borkdude/refl. It filters out many false positives from the agent with regards to Clojure.

**Reply by luontola** [reply]
@borkdude Now there is one more:  https://github.com/luontola/clojure-native-image-agent ✌️

It will initialize a Clojure app's main class and track all classes that the JVM loads during that. It also avoids native-image-agent reporting all that happened during Clojure namespace initialization. 

For now it prints the full list of classes that were loaded. The obvious next improvement is to deduct the package names for Clojure namespaces and generate a native-image.properties file. Java classes are probably best listed individually, in case the same package contains classes that don't support BTI.

**Reply by borkdude** [reply]
@luontola Isn't your project tracking all loaded classes instead of only the ones that are instantiated at build time transitively? All loaded classes is basically the same as a global `--initialize-at-build-time`. The `clj-easy/graal-build-time` leaves all Java classes alone that aren't used in top level expressions for example.

**Reply by luontola** [reply]
It tracks only those that are loaded during initialization. Not those that area loaded after the main method is called. native-image-agent will report those that are loaded after calling main.

**Reply by borkdude** [reply]
Ah, that's interesting, thanks, I'll have a closer look when I have time.

**Reply by luontola** [reply]
For example, I'm currently writing an AWS Lambda function. clojure-native-image-agent reports about 5000 classes. After ignoring all Clojure functions etc. it's about 1000 classes. The code uses AWS Java SDK, but the tool reports only these 10 classes. If it would actually report all the classes that the application uses at runtime, there would be hundreds of com.amazonaws classes.

```
com.amazonaws.AmazonWebServiceRequest
com.amazonaws.HandlerContextAware
com.amazonaws.ReadLimitInfo
com.amazonaws.client.builder.AwsClientBuilder
com.amazonaws.client.builder.AwsSyncClientBuilder
com.amazonaws.services.kms.AWSKMSClientBuilder
com.amazonaws.services.kms.model.DecryptRequest
com.amazonaws.services.lambda.runtime.Context
com.amazonaws.services.lambda.runtime.RequestStreamHandler
com.amazonaws.services.lambda.runtime.api.client.AWSLambda
```

**Reply by borkdude** [reply]
I made an issue [here](https://github.com/clj-easy/graal-build-time/issues/15) to investigate if we can combine both approaches:

- Configure during native image build (clj-easy/graal-build-time)
- Use Java agent (or another way) to detect transitively instantiated classes (luontola's approach)

One question that arises from this for @thomaswue: could GraalVM have an option to automatically include transitively instantiated classes at build time, in addition to the trace option? That would save us from building this tool in the first place. I guess the reason not to would be that it would be too easy to do this?

**Reply by vjovanov** [reply]
It would be convenient, however, this would lead to accidental errors as unknown classes get initialized. Also, this is not trivial to do as it is hard to know exactly what are the transitively instantiated classes.

**Reply by borkdude** [reply]
I agree the tool should emit warnings about classes that are being pulled in unexpectedly.

**Comment by fcurts** [comment]
> How are those libraries referenced, is it via maven artefacts?

Yes.

> Is there a way for those libraries to declare themselves as "build time initialization required"?

These are OSS libraries. It may make sense for these libraries to declare whether they want to be initialized at build time, but for our Truffle language I wouldn't want to rely on this. If there's no general wildcard, we'll probably have our build generate the list of packages.

> Does it also work with the automatic detection when nothing is specified

We've used `--initialize-at-build-time` from day one and haven't tried anything else. The most critical build time initialization is certainly for our own code (e.g., parse/initialize our Truffle language's standard library).
